### PR TITLE
Rdrp 541 postcodes

### PR DESCRIPTION
**Pull request by zorge69**
## Pull Request submission

*Harmonised postcodes have lengths from 5 to 9. Harmonised them so that they are always 8.

*Insert any instructions to help the reviewer, e.g. "install new requirements from `environment.yml`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/outputs/outputs_main.py [hunk 1]
```diff
@@ -82,6 +82,37 @@ def run_outputs(
...
+        if len(parts) >= 2:
+            part1 = parts[0]
+            part2 = parts[1]
+            num_spaces = target_len - len(part1) - len(part2)
+            if num_spaces >= 0:
+                return part1 + " " * num_spaces + part2
+            else:
+                return (part1 + part2)[:target_len]
+        else:
+            return mystr[:target_len].ljust(target_len, " ")
```
Comment (coatet) [inline src/outputs/outputs_main.py]: This function shouldn't be defined inside the `run_outputs()` function; should either be in a separate file or defined in this file outside the main function.

File: src/outputs/outputs_main.py [hunk 2]
```diff
@@ -82,6 +82,37 @@ def run_outputs(
...
+    # Topping up the postcodes
+    def _topup(mystr: str, target_len: int = 8):
+        ''' Regulates the number of spaces in a postcode, so that the total
+        length is 8 characters.
+        Brings all letters to upper case, in line with the mapper.
+        Splits it into parts, separated by any number of spaces.
+        If there are two or more parts, the first two are used.
+            If their total lenght is too big, cuts the tail
+        If it's only one part, keeps 8 characters, tops up with spaces if needed
+        '''
```
Comment (coatet) [inline src/outputs/outputs_main.py]: Docstring needs reformatting into the standard format.

File: pc_lenths_prototype.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+"""
+import pandas as pd
+
+mydata = {"pc": ["A B", "A B C", "1234 567", "12345678", "123456789"]}
+df = pd.DataFrame(mydata)
+
+
+target_len = 8
+_topup
+df['pc'] = df['pc'].apply(_topup)
```
Comment (coatet) [inline pc_lenths_prototype.py]: This file should be removed since it was just for testing.

**Review by coatet — CHANGES_REQUESTED** [review]

**Review by AnneONS — COMMENTED** [review]
Hi @zorge69 
Good that you managed to fix the bug! However, we're doing postcode validation in the staging module, so it would be good to have the fix before we do the postcode validation.

The functions Ilyas wrote to do this are in staging/validation.py. He has also written code that was supposed to change the patterns to 8 digits, and I thought it was working.

The ideal outcome would be if we could fix whatever bug was in there, but I realise we also need to work at pace.

File: pc_lenths_prototype.py (hunk 3 above)
Comment (zorge69) [reply pc_lenths_prototype.py]: removed

**Comment by zorge69** [comment]
Created a new bug [ticket|https://jira.ons.gov.uk/browse/RDRP-555] to move topping up to staging and validation. Not essential for data output, low priority.

**Comment by zorge69** [comment]
Resolved conflicts

File: src/outputs/outputs_main.py (hunk 1 above)
Comment (zorge69) [reply src/outputs/outputs_main.py]: Moved the function to output_helpers

File: src/outputs/outputs_main.py (hunk 2 above)
Comment (zorge69) [reply src/outputs/outputs_main.py]: Updated the docstring

**Review by coatet — APPROVED** [review]

**Comment by coatet** [comment]
Looks like when you merged in develop the import for the FTE QA output was lost in outputs_main, but I've fixed that and pushed it. Ready to merge!

**Pull request merged**